File: moderate/alternative_reality.pl
use strict;
my @m = ( 50, 25, 10, 5, 1 );

sub alter {
    my ( $n, $p ) = @_;
    return (1) if ( $n == 0 );
    $p += 1 while ( $m[$p] > $n );
    return (1) if ( $m[$p] == 1 );
    return ( alter( $n - $m[$p], $p ) + alter( $n, $p + 1 ) );
}
open( INFILE, "<", $ARGV[0] )
  or die("Cannot open file $ARGV[0] for reading: $!");
while ( my $line = <INFILE> ) {
    printf( "%d\n", alter( $line, 0 ) );
}
close(INFILE);

File: moderate/alternative_reality.rb
M = [50, 25, 10, 5, 1].freeze

def alter(n, p)
  return 1 if n == 0
  p += 1 while M[p] > n
  return 1 if M[p] == 1
  alter(n - M[p], p) + alter(n, p + 1)
end

File.open(ARGV[0]).each_line do |line|
  puts alter(line.to_i, 0)
end
